src/pm-conversations/pmConversationsSelectors.js

```diff
@@ -51,10 +67,114 @@ export const getRecentConversations: Selector<PmConversationData[]> = createSele
          end up converted to strings, so this access with string keys works.  We should probably use
          a Map for this and similar maps. */
         unreadPms[recipient.ids] || unreadHuddles[recipient.ids],
+    })),
+);
+
+/**
+ * Legacy implementation of {@link getRecentConversations}. Computes an
+ * approximation to the set of recent conversations, based on the messages we
+ * already know about.
+ */
+const getRecentConversationsLegacyImpl: Selector<PmConversationData[]> = createSelector(
+  getOwnUser,
+  getPrivateMessages,
+  getAttachUnread,
+  (ownUser: User, messages: Message[], attachUnread: UnreadAttacher): PmConversationData[] => {
+    const items = messages.map(msg => ({
+      ids: pmUnreadsKeyFromMessage(msg, ownUser.user_id),
+      recipients: normalizeRecipientsSansMe(msg.display_recipient, ownUser.email),
+      msgId: msg.id,
     }));
+
+    const sortedByMostRecent = collateByRecipient(items);
+
+    return attachUnread(sortedByMostRecent);
   },
 );
 
+/**
+ * Modern implementation of {@link getRecentConversations}. Returns exactly the
+ * most recent conversations. Requires server-side support.
+ */
+const getRecentConversationsImpl: Selector<PmConversationData[]> = createSelector(
+  getOwnUser,
+  getAllUsersById,
+  getRecentPrivateConversations,
+  getAttachUnread,
+  (
+    ownUser: User,
+    allUsers: Map<number, UserOrBot>,
+    recentPCs: RecentPrivateConversation[],
+    attachUnread: UnreadAttacher,
+  ) => {
+    const getEmail = (id: number): string => {
+      const user = allUsers.get(id);
+      if (user) {
+        return user.email;
+      }
+      throw new Error('getRecentConversations: unknown user id');
+    };
+
+    const recipients = recentPCs.map(conversation => {
+      const userIds = conversation.user_ids.slice();
+      if (userIds.length !== 1) {
+        userIds.push(ownUser.user_id);
+        userIds.sort((a, b) => a - b);
+      }
```

Comment:
This should also go in `recipients.js`, with a name and jsdoc.

Specifically, a good name/interface might be:
* include the `.join(',')` part;
* call it `pmUnreadsKeyFromRecentPmsKey`.